feat: Customize Gateway logs section (#880)
Kong/developer.konghq.com: app/gateway/logs.md

## Customize what {{site.base_gateway}} logs

> You might need to customize what {{site.base_gateway}} logs to protect private information and comply with GDPR. For example, if you wanted to remove instances of an email address from your logs. These changes can be made to {{site.base_gateway}}'s Nginx template and only affect the output of the Nginx access logs. This doesn't have any effect on {{site.base_gateway}}'s [logging plugins](/plugins/?category=logging).

Reviewer: ```suggestion
You may need to customize what {{site.base_gateway}} logs. For instance, you may want to:
* Protect private information
* Comply with GDPR or other data protection regulations
* Remove instances of a specific piece of data from your logs, such as an email address

These changes can be made to {{site.base_gateway}}'s Nginx template and only affect the output of the Nginx access logs. This doesn't have any effect on {{site.base_gateway}}'s [logging plugins](/plugins/?category=logging).
```

> For this example, let’s say you want to remove any instances of an email address from your {{site.base_gateway}} logs. The email addresses may come through in different ways, for example `/servicename/v2/verify/alice@example.com` or `/v3/verify?alice@example.com`. To keep all of these formats from being added to the logs, you need to use a custom Nginx template by making a copy of {{site.base_gateway}}'s Nginx template.

Reviewer: ```suggestion
For this example, let’s say you want to remove any instances of an email address from your {{site.base_gateway}} logs. The email addresses may come through in different formats, for example `/servicename/v2/verify/alice@example.com` or `/v3/verify?alice@example.com`. To keep all of these formats from being added to the logs, you need to use a custom Nginx template.
```

> The following template shows an example configuration to remove email addresses from logs, but you can use this logic to remove anything you want from the logs in a conditional manner:

Reviewer: ```suggestion
Make a copy of {{site.base_gateway}}'s Nginx template, then edit it to add or remove the data you need. The following template shows an example configuration for removing email addresses from logs:
```
I don't think you need the "you can use this same config for everything", that seems implied. Otherwise we'd put that in front of every example.

Reviewer: We should link to wherever this info lives now, so they can figure out how to find the template: https://docs.konghq.com/gateway/latest/reference/nginx-directives/#custom-nginx-templates-and-embedding-kong-gateway

```nginx
# ---------------------
# custom_nginx.template
# ---------------------

worker_processes ${{NGINX_WORKER_PROCESSES}}; # can be set by kong.conf
daemon ${{NGINX_DAEMON}};                     # can be set by kong.conf

pid pids/nginx.pid;                      # this setting is mandatory
error_log stderr ${{LOG_LEVEL}}; # can be set by kong.conf

events {
    use epoll; # custom setting
    multi_accept on;
}

http {

    map $request_uri $keeplog {
        ~.+\@.+\..+ 0;
        ~/v1/invitation/ 0;
        ~/reset/v1/customer/password/token 0;
        ~/v2/verify 0;

        default 1;
    }
    log_format show_everything '$remote_addr - $remote_user [$time_local] '
        '$request_uri $status $body_bytes_sent '
        '"$http_referer" "$http_user_agent"';

    include 'nginx-kong.conf';
}
```

For this example, we're using the following:

> * `map $request_uri $keeplog`: Maps a new variable called `keeplog`, which is dependent on values appearing in the `$request_uri`. Each line in the example starts with a `~` because this is what tells Nginx to use regex when evaluating the line. This example looks for the following:

Reviewer: ```suggestion
* `map $request_uri $keeplog`: Maps a new variable called `keeplog`, which is dependent on values appearing in the `$request_uri`. Each line in the example starts with a `~` because this is what tells Nginx to use a regex when evaluating the line. This example looks for the following:
```

>   - The first line uses regex to look for any email address in the `x@y.z` format

Reviewer: ```suggestion
  - The first line uses a regex to look for any email address in the `x@y.z` format
```

  - The second line looks for any part of the URI that contains `/servicename/v2/verify`
  - The third line looks at any part of the URI that contains `/v3/verify`
  Because all of these patterns have a value of something other than `0`, if a request has any of those elements, it will not be added to the log.

> * `log_format`: Sets the log format for what {{site.base_gateway}} keeps in the logs. The contents of the log can be customized for your needs. For the purpose of this example, you can assign the new logs with the name `show_everything` and  simply set everything to the {{site.base_gateway}} default standards. To see the full list of options, refer to the [Nginx core module variables reference](https://nginx.org/en/docs/http/ngx_http_core_module.html#variables).

Reviewer: ```suggestion
* `log_format`: Sets the log format for what {{site.base_gateway}} keeps in the logs. The contents of the log can be customized for your needs. For the purpose of this example, you can assign the new logs with the name `show_everything` and simply set everything to the {{site.base_gateway}} default standards. To see the full list of options, refer to the [Nginx core module variables reference](https://nginx.org/en/docs/http/ngx_http_core_module.html#variables).
```

## Customize what {{site.base_gateway}} logs

You may need to customize what {{site.base_gateway}} logs. For instance, you may want to:
* Protect private information
* Comply with GDPR or other data protection regulations
* Remove instances of a specific piece of data from your logs, such as an email address

These changes can be made to {{site.base_gateway}}'s Nginx template and only affect the output of the Nginx access logs. This doesn't have any effect on {{site.base_gateway}}'s [logging plugins](/plugins/?category=logging).

> For this example, let’s say you want to remove any instances of an email address from your {{site.base_gateway}} logs. The email addresses may come through in different formats, for example `/servicename/v2/verify/alice@example.com` or `/v3/verify?alice@example.com`. To keep all of these formats from being added to the logs, you need to use a custom Nginx template.

Reviewer: ```suggestion
Let's look at an example where you want to remove any instances of an email address from your {{site.base_gateway}} logs. The email addresses may come through in different formats, for example `/servicename/v2/verify/alice@example.com` or `/v3/verify?alice@example.com`. To keep all of these formats from being added to the logs, you need to use a custom Nginx template.
```
Rephrasing for better flow with the previous paragraphs.

Make a copy of [{{site.base_gateway}}'s Nginx template](/gateway/nginx-directives/#custom-nginx-templates-and-embedding-kong-gateway), then edit it to add or remove the data you need. The following template shows an example configuration for removing email addresses from logs:

```nginx
# ---------------------
# custom_nginx.template
# ---------------------

worker_processes ${{NGINX_WORKER_PROCESSES}}; # can be set by kong.conf
daemon ${{NGINX_DAEMON}};                     # can be set by kong.conf

pid pids/nginx.pid;                      # this setting is mandatory
error_log stderr ${{LOG_LEVEL}}; # can be set by kong.conf

events {
    use epoll; # custom setting
    multi_accept on;
}

http {

    map $request_uri $keeplog {
        ~.+\@.+\..+ 0;
        ~/v1/invitation/ 0;
        ~/reset/v1/customer/password/token 0;
        ~/v2/verify 0;

        default 1;
    }
    log_format show_everything '$remote_addr - $remote_user [$time_local] '
        '$request_uri $status $body_bytes_sent '
        '"$http_referer" "$http_user_agent"';

    include 'nginx-kong.conf';
}
```

For this example, we're using the following:

> * `map $request_uri $keeplog`: Maps a new variable called `keeplog`, which is dependent on values appearing in the `$request_uri`. Each line in the example starts with a `~` because this is what tells Nginx to use a regex when evaluating the line. This example looks for the following:
>   - The first line uses a regex to look for any email address in the `x@y.z` format
>   - The second line looks for any part of the URI that contains `/servicename/v2/verify`
>   - The third line looks at any part of the URI that contains `/v3/verify`
>   Because all of these patterns have a value of something other than `0`, if a request has any of those elements, it will not be added to the log.

Reviewer: ```suggestion
* `map $request_uri $keeplog`: Maps a new variable called `keeplog`, which is dependent on values appearing in the `$request_uri`. Each line in the example starts with a `~` because this is what tells Nginx to use a regex when evaluating the line. This example looks for the following:
  - The first line uses a regex to look for any email address in the `x@y.z` format
  - The second line looks for any part of the URI that contains `/servicename/v2/verify`
  - The third line looks at any part of the URI that contains `/v3/verify`
    
    Because all of these patterns have a value of something other than `0`, if a request has any of those elements, it will not be added to the log.
```
I think this is the right indent, but not 100% sure. Right now this sentence is getting added to the previous line.

* `log_format`: Sets the log format for what {{site.base_gateway}} keeps in the logs. The contents of the log can be customized for your needs. For the purpose of this example, you can assign the new logs with the name `show_everything` and set everything to the {{site.base_gateway}} default standards. To see the full list of options, refer to the [Nginx core module variables reference](https://nginx.org/en/docs/http/ngx_http_core_module.html#variables).

> Once you've adjusted the Nginx template for your environment, the last thing you must do is tell {{site.base_gateway}} to use the newly created log, `show_everything`. 

Reviewer: ```suggestion
Once you've adjusted the Nginx template for your environment, you need to tell {{site.base_gateway}} to use the newly created log, `show_everything`. 
```

To do this, alter the {{site.base_gateway}} variable `proxy_access_log` by either editing `etc/kong/kong.conf` or using the environmental variable `KONG_PROXY_ACCESS_LOG` adjust the default location:

```sh
proxy_access_log=logs/access.log show_everything if=$keeplog
```

> Restart {{site.base_gateway}} to make all the changes take effect with the `kong restart` command.

Reviewer: ```suggestion
Restart {{site.base_gateway}} to apply changes with the `kong restart` command.
```
